### PR TITLE
Collapse worker select strategies, change default, add strong affinity.

### PR DESCRIPTION
**Pull request by gianm**
- Change default worker select strategy to equalDistribution. It is
  more generally useful than fillCapacity.
- Collapse the *WithAffinity strategies into the regular ones. The
  *WithAffinity strategies are retained for backwards compatibility.
- Change WorkerSelectStrategy to return nullable instead of Optional.
- Fix a couple of errors in the docs.

The motivation is to simplify the worker select config and to make the builtin ones more useful in more situations.

Marked release notes since the default worker select strategy would change.

### CODE REVIEW

**Comment by fjy** [comment]
👍

File: indexing-service/src/main/java/io/druid/indexing/overlord/setup/WorkerSelectUtils.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+          .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+      // Workers assigned to the affinity pool for our task.
+      final List<String> dataSourceWorkers = affinityConfig.getAffinity().get(task.getDataSource());
+
+      if (dataSourceWorkers == null) {
+        // No affinity config for this dataSource; use non-affinity workers.
+        return workerSelector.apply(ImmutableMap.copyOf(nonAffinityWorkers));
+      } else {
+        final Set<String> dataSourceWorkerHosts = ImmutableSet.copyOf(dataSourceWorkers);
```
Comment (jihoonson) [inline indexing-service/src/main/java/io/druid/indexing/overlord/setup/WorkerSelectUtils.java]: Maybe don't have to copy if ```AffinityConfig.affinity``` is a ```Map<String, Set<String>>```?

File: indexing-service/src/main/java/io/druid/indexing/overlord/setup/FillCapacityWorkerSelectStrategy.java [hunk 2]
```diff
@@ -19,59 +19,87 @@
...
+        FillCapacityWorkerSelectStrategy::selectFromEligibleWorkers
+    );
+  }
 
-            return retVal;
-          }
-        }
+  private static ImmutableWorkerInfo selectFromEligibleWorkers(final Map<String, ImmutableWorkerInfo> eligibleWorkers)
+  {
+    TreeSet<ImmutableWorkerInfo> sortedWorkers = Sets.newTreeSet(
```
Comment (jihoonson) [inline indexing-service/src/main/java/io/druid/indexing/overlord/setup/FillCapacityWorkerSelectStrategy.java]: Does this TreeSet is needed? It looks to simply find the worker of the max used capacity.

File: indexing-service/src/main/java/io/druid/indexing/overlord/setup/EqualDistributionWorkerSelectStrategy.java [hunk 3]
```diff
@@ -19,41 +19,89 @@
...
         Comparator.comparing(ImmutableWorkerInfo::getAvailableCapacity).reversed()
-                  .thenComparing(zkWorker -> zkWorker.getWorker().getVersion()));
-    sortedWorkers.addAll(zkWorkers.values());
-    final String minWorkerVer = config.getMinWorkerVersion();
-
-    for (ImmutableWorkerInfo zkWorker : sortedWorkers) {
-      if (zkWorker.canRunTask(task) && zkWorker.isValidVersion(minWorkerVer)) {
-        return Optional.of(zkWorker);
-      }
+    );
```
Comment (jihoonson) [inline indexing-service/src/main/java/io/druid/indexing/overlord/setup/EqualDistributionWorkerSelectStrategy.java]: Same here. Does this TreeSet is needed?

File: indexing-service/src/main/java/io/druid/indexing/overlord/setup/FillCapacityWorkerSelectStrategy.java (hunk 2 above)
Comment (gianm) [reply indexing-service/src/main/java/io/druid/indexing/overlord/setup/FillCapacityWorkerSelectStrategy.java]: That's a good point. I'll change it to be a simpler `min` function.

File: indexing-service/src/main/java/io/druid/indexing/overlord/setup/EqualDistributionWorkerSelectStrategy.java (hunk 3 above)
Comment (gianm) [reply indexing-service/src/main/java/io/druid/indexing/overlord/setup/EqualDistributionWorkerSelectStrategy.java]: Will also change this to a simpler `min` function.

File: indexing-service/src/main/java/io/druid/indexing/overlord/setup/WorkerSelectUtils.java (hunk 1 above)
Comment (gianm) [reply indexing-service/src/main/java/io/druid/indexing/overlord/setup/WorkerSelectUtils.java]: Good point, I'll change it.

File: indexing-service/src/main/java/io/druid/indexing/overlord/setup/WorkerSelectUtils.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+      if (dataSourceWorkers == null) {
+        // No affinity config for this dataSource; use non-affinity workers.
+        return workerSelector.apply(ImmutableMap.copyOf(nonAffinityWorkers));
+      } else {
+        final Map<String, ImmutableWorkerInfo> dataSourceWorkerMap = runnableWorkers
+            .entrySet()
+            .stream()
+            .filter(entry -> dataSourceWorkers.contains(entry.getKey()))
+            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
```
Comment (egor-ryashin) [inline indexing-service/src/main/java/io/druid/indexing/overlord/setup/WorkerSelectUtils.java]: Could be as concise as, i.e. `Map<String, ImmutableWorkerInfo> filtered = Maps.filterKeys(runnableWorkers, s -> dataSourceWorkers.contains(s));`

See `com.google.common.collect.Maps`

File: indexing-service/src/main/java/io/druid/indexing/overlord/setup/WorkerSelectUtils.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    if (affinityConfig == null) {
+      return workerSelector.apply(ImmutableMap.copyOf(runnableWorkers));
+    } else {
+      // Workers not assigned to any affinity pool at all.
+      final Map<String, ImmutableWorkerInfo> nonAffinityWorkers = runnableWorkers
+          .entrySet()
+          .stream()
+          .filter(entry -> !affinityConfig.getAffinityWorkers().contains(entry.getKey()))
+          .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
```
Comment (egor-ryashin) [inline indexing-service/src/main/java/io/druid/indexing/overlord/setup/WorkerSelectUtils.java]: Could be more compact, see a comment below.

File: indexing-service/src/main/java/io/druid/indexing/overlord/setup/WorkerSelectUtils.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+        final ImmutableWorkerInfo selected = workerSelector.apply(ImmutableMap.copyOf(dataSourceWorkerMap));
+
+        if (selected != null) {
+          return selected;
+        } else if (affinityConfig.isStrong()) {
+          return null;
+        } else {
+          // Weak affinity allows us to use nonAffinityWorkers for this dataSource, if no affinity workesr
```
Comment (egor-ryashin) [inline indexing-service/src/main/java/io/druid/indexing/overlord/setup/WorkerSelectUtils.java]: typo

File: indexing-service/src/main/java/io/druid/indexing/overlord/setup/WorkerSelectUtils.java [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+        .filter(worker -> worker.canRunTask(task)
+                          && worker.isValidVersion(workerTaskRunnerConfig.getMinWorkerVersion()))
+        .collect(Collectors.toMap(w -> w.getWorker().getHost(), Function.identity()));
+
+    if (affinityConfig == null) {
+      return workerSelector.apply(ImmutableMap.copyOf(runnableWorkers));
+    } else {
+      // Workers not assigned to any affinity pool at all.
+      final Map<String, ImmutableWorkerInfo> nonAffinityWorkers = runnableWorkers
+          .entrySet()
```
Comment (egor-ryashin) [inline indexing-service/src/main/java/io/druid/indexing/overlord/setup/WorkerSelectUtils.java]: `nonAffinityWorkers` is not used in every `if` branch, could you extract a method and call it where it is required.

File: indexing-service/src/main/java/io/druid/indexing/overlord/setup/FillCapacityWorkerSelectStrategy.java [hunk 8]
```diff
@@ -19,59 +19,82 @@
...
-    final String minWorkerVer = config.getMinWorkerVersion();
+  }
 
-    for (ImmutableWorkerInfo zkWorker : sortedWorkers) {
-      if (zkWorker.canRunTask(task) && zkWorker.isValidVersion(minWorkerVer)) {
-        return Optional.of(zkWorker);
-      }
+  private static ImmutableWorkerInfo selectFromEligibleWorkers(final Map<String, ImmutableWorkerInfo> eligibleWorkers)
+  {
+    return eligibleWorkers.values().stream().min(
```
[the remark this answers is not in the document]
Comment (gianm) [reply indexing-service/src/main/java/io/druid/indexing/overlord/setup/FillCapacityWorkerSelectStrategy.java]: Sure, that's simpler, I'll change it.

File: indexing-service/src/main/java/io/druid/indexing/overlord/setup/EqualDistributionWorkerSelectStrategy.java [hunk 9]
```diff
@@ -19,41 +19,84 @@
...
         Comparator.comparing(ImmutableWorkerInfo::getAvailableCapacity).reversed()
-                  .thenComparing(zkWorker -> zkWorker.getWorker().getVersion()));
-    sortedWorkers.addAll(zkWorkers.values());
-    final String minWorkerVer = config.getMinWorkerVersion();
-
-    for (ImmutableWorkerInfo zkWorker : sortedWorkers) {
-      if (zkWorker.canRunTask(task) && zkWorker.isValidVersion(minWorkerVer)) {
-        return Optional.of(zkWorker);
-      }
+    ).orElse(null);
```
[the remark this answers is not in the document]
Comment (gianm) [reply indexing-service/src/main/java/io/druid/indexing/overlord/setup/EqualDistributionWorkerSelectStrategy.java]: Sure, it's simpler, I'll change it.

File: indexing-service/src/main/java/io/druid/indexing/overlord/setup/WorkerSelectUtils.java (hunk 7 above)
Comment (gianm) [reply indexing-service/src/main/java/io/druid/indexing/overlord/setup/WorkerSelectUtils.java]: OK.

File: indexing-service/src/main/java/io/druid/indexing/overlord/setup/WorkerSelectUtils.java (hunk 4 above)
Comment (gianm) [reply indexing-service/src/main/java/io/druid/indexing/overlord/setup/WorkerSelectUtils.java]: Changed it.

File: indexing-service/src/main/java/io/druid/indexing/overlord/setup/WorkerSelectUtils.java (hunk 5 above)
Comment (gianm) [reply indexing-service/src/main/java/io/druid/indexing/overlord/setup/WorkerSelectUtils.java]: Changed it.

File: indexing-service/src/main/java/io/druid/indexing/overlord/setup/WorkerSelectUtils.java (hunk 6 above)
Comment (gianm) [reply indexing-service/src/main/java/io/druid/indexing/overlord/setup/WorkerSelectUtils.java]: Fixed.

**Comment by gianm** [comment]
@egor-ryashin thanks for your review; I pushed some new code.

**Pull request merged**